src/main/js/utils/generate.js

```diff
@@ -14,50 +15,55 @@ const generateSchedules = (values, loadQuery) => (dispatch, getState) => {
   const { courseList } = getState().courses;
   if (courseList.length !== 0) {
     dispatch(startGenerating());
-    const filteredResults = filterCourses(values, courseList);
-    dispatch(filteredCourses(filteredResults));
-    const filteredCourseList = filteredResults.map(result => (
-      result.filtered));
-    let schedules = [];
-    if (!filteredCourseList.find(list => list.length === 0)) {
-      const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
-      try {
-        schedules = scheduleMaker.makeSchedules();
-      } catch (e) {
-        schedules = e;
+    addGPA(courseList).then((res) => {
+      const filteredResults = filterCourses(values, res);
+      dispatch(filteredCourses(filteredResults));
+      const filteredCourseList = filteredResults.map(result => (
+        result.filtered));
+      let schedules = [];
+      if (!filteredCourseList.find(list => list.length === 0)) {
+        const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
+        try {
+          schedules = scheduleMaker.makeSchedules();
+        } catch (e) {
+          schedules = e;
+        }
+        if (values.sortByGPA === 'yes') {
+          schedules.sort((schedule1, schedule2) => schedule2.gpa - schedule1.gpa);
```

Comment:
Didn't we talk about sorting the courseList by gpa before sending it to the ScheduleMaker so that it comes out sorted?

Reply:
We did, and after some testing it turns out that my theory was right. Because Schedule GPAs are also influenced by the credit hour of a class, then there are ways to obtain a higher Schedule GPA from midway into the createSchedule() recursion, rather than the first couple ones using the highest course GPA's.

Reply:
If each each list within courseLists have all the same credit hours how is that possible? Can you provide an example?

Reply:
Assuming my thinking is correct in the above comments:

Still, sometimes all the courses within a courseList will not have the same credit hours. But once we generate the first set of schedules, we can say that the resulting schedules are _mostly_ sorted. This means that we can use a sorting algorithm that runs efficiently on _mostly_ sorted lists (order n). Javascript's Array.sort happens to use [Timsort](https://v8.dev/blog/array-sort) which is specialized for this [use case](https://stackoverflow.com/a/220750).

Thus I think the best efficiency we can get is to:

Sort each of the courseLists by GPA.
O(nlogn), n being the size of the largest courseList.

Then running `schedules.sort`
O(n), n being the number of schedules generated.

Reply:
Used Inserstion sort to sort the mostly sorted array of schedules.